[WPE][GTK] Release crash in WebCore::Region

philn opened the pull request:
#### 57fb82f2bec55d418de1a2f43cfc29840a7d3cdb
<pre>
[WPE][GTK] Release crash in WebCore::Region
<a href="https://bugs.webkit.org/show_bug.cgi?id=292098">https://bugs.webkit.org/show_bug.cgi?id=292098</a>

Reviewed by Carlos Garcia Campos.

Fix a crash in Damage, there is no Region::unite() method accepting a Rect, so make a
Region and then unite it with the main region.

* Source/WebCore/platform/graphics/Damage.h:
(WebCore::FrameDamageHistory::addDamage):

Canonical link: <a href="https://commits.webkit.org/294501@main">https://commits.webkit.org/294501@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c5acc614da9b6b4279b94b75334acd0231cdd13

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101965 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21633 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11949 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107124 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52599 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104005 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21941 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30140 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77626 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34631 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104972 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16961 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92063 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57962 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16788 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10088 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51957 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86628 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10161 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109494 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29098 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21432 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86600 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29459 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88267 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86179 "Found 102 new API test failures: /WebKitGTK/TestUIClient:/webkit/WebKitWebView/javascript-dialogs, /TestWebKit:WebKit.DOMWindowExtensionCrashOnReload, /TestWebKit:WebKit.PageLoadBasic, /TestWebKit:WebKit.NewFirstVisuallyNonEmptyLayoutFails, /WebKitGTK/TestInputMethodContext:/webkit/WebKitInputMethodContext/content-type, /TestWebKit:WebKit.InjectedBundleBasic, /TestWebKit:WebKit.PreventEmptyUserAgent, /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/autoplay-policy, /WebKitGTK/TestInputMethodContext:/webkit/WebKitInputMethodContext/surrounding, /WebKitGTK/TestWebViewEditor:/webkit/WebKitWebView/select-all/non-editable ... (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21938 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30950 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8668 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23284 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29026 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34321 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28837 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32160 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30396 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -361,8 +361,10 @@ class FrameDamageHistory {
+        for (const auto& rect : damage.rects()) {
+            Region subRegion(rect);
+            region.unite(subRegion);
+        }
```
I don't understand the crash yet, but maybe after https://github.com/WebKit/WebKit/pull/43892 we no longer need to use Region for this.

philn replied:
Still happening in main:

```
Thread 1 (Thread 0x7fc8a5fff6c0 (LWP 2440443)):
#0  0x00007fca3fdf0373 in WebCore::Region::Region(WebCore::Region&&) () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libwebkitgtk-6.0.so.4
#1  0x00007fca3dff8ec8 in WebCore::FrameDamageHistory::addDamage(WebCore::Damage const&) () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libwebkitgtk-6.0.so.4
#2  0x00007fca3dff8a8d in WebKit::ThreadedCompositor::paintToCurrentGLContext(WebCore::TransformationMatrix const&, WebCore::IntSize const&) () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libwebkitgtk-6.0.so.4
#3  0x00007fca3dff915c in WebKit::ThreadedCompositor::renderLayerTree() () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libwebkitgtk-6.0.so.4
#4  0x00007fca3a1ceb0b in WTF::RunLoop::TimerBase::TimerBase(WTF::Ref<WTF::RunLoop, WTF::RawPtrTraits<WTF::RunLoop>, WTF::DefaultRefDerefTraits<WTF::RunLoop> >&&)::$_0::__invoke(void*) () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libjavascriptcoregtk-6.0.so.1
#5  0x00007fca3a1cde8a in WTF::RunLoop::$_0::__invoke(_GSource*, int (*)(void*), void*) () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libjavascriptcoregtk-6.0.so.1
#6  0x00007fca35d38040 in g_main_dispatch (context=0x7fc89c000b70) at ../glib/gmain.c:3398
#7  g_main_context_dispatch_unlocked (context=0x7fc89c000b70) at ../glib/gmain.c:4249
#8  0x00007fca35d41128 in g_main_context_iterate_unlocked (context=0x7fc89c000b70, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at ../glib/gmain.c:4314
#9  0x00007fca35d413cf in g_main_loop_run (loop=0x7fc89c000db0) at ../glib/gmain.c:4516
#10 0x00007fca3a1ce32c in WTF::RunLoop::run() () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libjavascriptcoregtk-6.0.so.1
#11 0x00007fca3a1692e4 in WTF::Thread::entryPoint(WTF::Thread::NewThreadContext*) () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libjavascriptcoregtk-6.0.so.1
#12 0x00007fca3a1d1f36 in WTF::wtfThreadEntryPoint(void*) () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libjavascriptcoregtk-6.0.so.1
#13 0x00007fca356f1f14 in start_thread (arg=<optimized out>) at pthread_create.c:448
#14 0x00007fca35774aac in __GI___clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:78
```

philn committed:
[WPE][GTK] Release crash in WebCore::Region
https://bugs.webkit.org/show_bug.cgi?id=292098

Reviewed by Carlos Garcia Campos.

Fix a crash in Damage, there is no Region::unite() method accepting a Rect, so make a
Region and then unite it with the main region.

* Source/WebCore/platform/graphics/Damage.h:
(WebCore::FrameDamageHistory::addDamage):

Canonical link: https://commits.webkit.org/294501@main
